hogben/simulate.py

```diff
@@ -10,368 +10,270 @@
+        if len(q) == 0:
+            return np.array([])
+
+        # Calculate the reflectance in either refnx or Refl1D.
+        if isinstance(self.sample_model, refnx.reflect.Stucture):
+            self.model = refnx.reflect.ReflectModel(self.sample_model,
+                                                    scale=self.scale,
+                                                    bkg=self.bkg,
+                                                    dq=self.dq)
+            return self.model(q)
+
+        if isinstance(self.sample_model, refl1d.model.Stack):
+            # If magnetic, use the correct spin state.
+            experiment = self.refl1d_experiment(q,
+                                                self.sample_model.probe.spin_state)
+
+            if self.sample_model.ismagnetic:
+                return experiment.reflectivity()[self.sample_model.probe.spin_state][1]
+            # experiment.reflectivity() returns q, r, or an array if magnetic
+
+            else:
+                return experiment.reflectivity()[1]
+
+    def _run_experiment(self, angle: float, points: int, time: float,
+                        spin_state: Optional[str] = None) -> tuple:
+        """Simulates a single angle measurement of a given 'model' on the
+        instrument set in self.incident_flux_data with the defined spin state,
+        unpolarised measurements (the default) have the set to None
+
+        Args:
+            angle: angle to simulate.
+            points: number of points to use for simulated data.
+            time: counting time for simulation.
+            spin_state: spin state to simulate if given a magnetic sample.
+
+        Returns:
+            tuple: simulated Q, R, dR data and incident neutron counts.
+
+        """
+        # direct_beam = [wavelength, flux]
+        wavelengths, flux = self.incident_flux_data
+
+        # Scale flux by relative measurement angle squared (assuming both slits
+        # scale linearly with angle, this should be correct)
+        scaled_flux = flux * pow(angle / self.angle_scale, 2)
+
+        q = 4 * np.pi * np.sin(np.radians(angle)) / wavelengths
+
+        # Bin Q's' in equally geometrically-spaced bins using flux as weighting
+        q_bin_edges = np.geomspace(min(q), max(q), points + 1)
+        flux_binned, _ = np.histogram(q, q_bin_edges, weights=scaled_flux)
+        # Calculate the number of incident neutrons for each bin.
+        counts_incident = flux_binned * time
+
+        # Get the bin centres.
+        q_binned = np.asarray(
+            [(q_bin_edges[i] + q_bin_edges[i + 1]) / 2 for i in range(points)])
+
+        # Calculate the model reflectivity at each Q point.
+        if isinstance(self.sample_model, refnx.reflect.Structure):
+            r_model = self.model(q_binned)
+
+        elif isinstance(self.sample_model, refl1d.model.Stack):
+            # TODO: remove instance checks if possible
+            r_model = self.reflectivity(q_binned, self.model)
```

Comment:
self.model is not defined upon init, instead it's only defined in simulate, and only for the refnx case. The refl1d case (this line) will thus result in an error.

Is it good practice to define new class attributes in lone methods? This may lead to issues, even in the refnx case, if we ever had a function that would run `_run_experiment` without running `simulate` first. Perhaps it's best practice to at least set place-holder attributes in the init first? 

I mean that as an open question, honestly not sure what is best. It just feels like it may be asking for trouble if we depend on certain methods being run first (in this case, `_run_experiment` only works if simulate has been run, so that `self.model` is defined), without explicitly doing so upon init or in this method here.

Reply:
Indeed, this does lead to an error. Perhaps `self.model` should be defined upon init?